server/src/main/java/org/opensearch/gateway/PrimaryShardAllocator.java

```diff
@@ -483,13 +537,126 @@ static class NodesToAllocate {
+        // The replication checkpoint of the shard
+        private final ReplicationCheckpoint replicationCheckpoint;
+        // The DiscoveryNode the shard belongs to
+        private final DiscoveryNode node;
+
+        /**
+         * Constructs a new NodeShardState with the given parameters.
+         * @param node The DiscoveryNode the shard belongs to.
+         * @param allocationId The allocation ID of the shard.
+         * @param primary Whether the shard is a primary shard.
+         * @param replicationCheckpoint The replication checkpoint of the shard.
+         * @param storeException Any store exception associated with the shard.
+         */
+        public NodeShardState(
+            DiscoveryNode node,
+            String allocationId,
+            boolean primary,
+            ReplicationCheckpoint replicationCheckpoint,
+            Exception storeException
+        ) {
+            this.node = node;
+            this.allocationId = allocationId;
+            this.primary = primary;
+            this.replicationCheckpoint = replicationCheckpoint;
+            this.storeException = storeException;
+        }
+
+        /**
+         * Returns the allocation ID of the shard.
+         * @return The allocation ID of the shard.
+         */
+        public String allocationId() {
+            return this.allocationId;
+        }
+
+        /**
+         * Returns whether the shard is a primary shard.
+         * @return True if the shard is a primary shard, false otherwise.
+         */
+        public boolean primary() {
+            return this.primary;
+        }
+
+        /**
+         * Returns the replication checkpoint of the shard.
+         * @return The replication checkpoint of the shard.
+         */
+        public ReplicationCheckpoint replicationCheckpoint() {
+            return this.replicationCheckpoint;
+        }
+
+        /**
+         * Returns any store exception associated with the shard.
+         * @return The store exception associated with the shard, or null if there isn't one.
+         */
+        public Exception storeException() {
+            return this.storeException;
+        }
+
+        /**
+         * Returns the DiscoveryNode the shard belongs to.
+         * @return The DiscoveryNode the shard belongs to.
+         */
+        public DiscoveryNode getNode() {
+            return this.node;
+        }
+    }
+
+    /**
+     * The NodeShardStates class manages pairs of {@link NodeShardState} and {@link DiscoveryNode}.
+     * It uses a TreeMap to ensure that the entries are sorted based on the natural
+     * ordering of the {@link NodeShardState} keys, or according to a provided Comparator.
+     * <p>
+     * The TreeMap is implemented using a Red-Black tree, which provides efficient
+     * performance for common operations such as adding, removing, and retrieving
+     * elements.
+     * @see TreeMap
+     */
+    protected static class NodeShardStates {
```

Comment:
We don't need this class. We can directly use the list of NodeShardState.

Reply:
Removed this class